src/services/organizeImports.ts

```diff
@@ -88,6 +87,44 @@ namespace ts.OrganizeImports {
         }
     }
 
+    function groupImportsByNewlineContiguous(sourceFile: SourceFile,importDecls: ImportDeclaration[]): ImportDeclaration[][] {
+        const groupImports: ImportDeclaration[][] = [];
+        let groupIndex = 0;
+        for (const topLevelImportDecl of importDecls) {
+            if (isStartedNewGroup(sourceFile, topLevelImportDecl)) {
+                groupIndex++;
+            }
+
+            if (!groupImports[groupIndex]) {
+                groupImports[groupIndex] = [];
+            }
+
+            groupImports[groupIndex].push(topLevelImportDecl);
+        }
+
+        return groupImports;
+    }
+
+    // a new group is created if an import includes at least two new line
+    // new line from multi-line comment doesn't count
+    function isStartedNewGroup(sourceFile: SourceFile, topLevelImportDecl: ImportDeclaration) {
+        const startPos = topLevelImportDecl.getFullStart();
+        const endPos = topLevelImportDecl.getStart();
+
+        const scanner = createScanner(sourceFile.languageVersion, /*skipTrivia*/ false, sourceFile.languageVariant, sourceFile.text, /*onError:*/ undefined, startPos);

```

Comment:
This feels like it's going to be expensive; is this not calculatable without rescanning? I would think that the node has on its trivia the preceding whitespace to check.

Reply:
thank for your suggestion, I move scanner out of that function to share between. That text can contain newlines and comments, I think in both ways we have to scan them to check 🤔

Reply:
I mean, I guess the formatter may do this when it's reprinting the tree, but it doesn't feel necessary in this particular case where we just want to look at how many newlines preceded the current import.

Is this a case where you're trying to go backwards in the tree to find the previous node? If so, there's a helper for that.

Reply:
hi @jakebailey,  I think comments and newlines are a part of node, in this case `topLevelImportDecl` like the image blow
![image](https://user-images.githubusercontent.com/1636026/161646589-e6d830d5-73ac-4a7e-a5e4-868e3e93aab2.png)

If I understand correctly, when emitting comments via `emitter.ts`, in some case we still rescan the text
![image](https://user-images.githubusercontent.com/1636026/161647552-8afeda49-beb9-400a-86aa-05d5f07dc64b.png)

In case I misunderstand, do you know which place to look at the going backwards 😄

Reply:
If I understand correctly, when emitting comments via `emitter.ts`, in some case we rescan the text
![image](https://user-images.githubusercontent.com/1636026/161647552-8afeda49-beb9-400a-86aa-05d5f07dc64b.png)

If I misunderstand, do you know which place to look at the going previous nodes (include comments) 😄. I think we have to traverse back for each node since counting only newline (regex text) is not safe since multlines comment can contains newline too
```
/**
 * hello world
 */
import { foo } from "bar"
```

Reply:
ah, I could misunderstand your thought when mentioning comments. In this case we want `path` and `anotherpath` belongs to different group, right?
```
import * as something from "path"; /**
 * some comment here
 * and there
 */
import * as somethingElse from "anotherpath";
```
I push changes 😄

Reply:
I'm not sure what example you're talking about here, but I think the "it's a new group" case I'm looking for here is if the trivia matches the regex `/\r?\n\s*\r?\n/g`, i.e. it contains two newlines in a row with maybe spaces between. Comments will always begin with something like `/*`, `*`, `*/`, or `//`, so will not match that regex and won't break the group.

Looking at the change you just made, I don't think it's correct in that it believes that the two imports `something` and `somethingElse` are in two different groups, when by the rule above, they aren't.

Reply:
I don't think it will work for the case like below?
```js
/**
 * some comment here
 * and there

asdasd asdasd
 */
```

Reply:
I want to clarify to make sure that we are on the same page, there two things to consider a new group:
- two consecutive newlines.
```js
import a from "a";

import b from "b";
```
- new line + comments (single line or multiline) + newline
```js
import a from "a";
// something here
import b from "b";
```

Reply:
Oh, you're right, it doesn't need a leading `*` at all, which breaks my rule. I was fooled by the test case that had it all `*`'d out.

I guess we'll have to scan after all. In that case, I guess the best way is to create the scanner over the trivia only, then scan; if two newline tokens are seen, then it's a new group, otherwise the scanner will end and it's not a new group.

Reply:
I reverted back and make changes to only scan trivia only 😄